### PR TITLE
Setup CD using GitHub actions

### LINKED ISSUE #121: Set up continuous deployment
Currently Vocabulary is being manually pushed to `npm`. This can be automated with CircleCI so that it can automatically build and publish the package once all tests pass successfully.

Best way to do this is to use some mechanism on Github Actions to detect when a new *tag* is pushed and trigger the run the [deploy script](https://github.com/creativecommons/vocabulary/blob/master/deploy.sh). It can't be the release name. It has to be able to detect a new tag, because the tag name matches the version on package.json.

### PR DESCRIPTION
**Pull request by SaurabhAgarwala**
## Fixes

Fixes #121 by @dhruvkb 

## Description
This PR adds a new workflow for Continuous Deployment using Github Actions which will push the updated packages to npm when a new release is published.

## Technical details

Please note that the NPM Authentication Token has to be stored as a GitHub secret by the name `NPM_TOKEN` for this workflow to complete successfully

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by dhruvkb — CHANGES_REQUESTED** [review]
We don't need to repeat the CI workflow here. The setup, build and deploy jobs should suffice.

**Comment by dhruvkb** [comment]
@kgodey could you [add the secret](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#creating-encrypted-secrets) for `NPM_TOKEN` in Settings for all three repos?

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+name: vocabulary-cd
+
+on:
+  release:
+    types: [published]
```
Comment (dhruvkb) [inline .github/workflows/cd.yml]: Lists are specified with `-` in YML

Comment (dhruvkb) [inline .github/workflows/cd.yml]:
```suggestion
    types: 
    - published
```

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+      - id: cache-node-modules
+        name: Cache Node.js modules
+        uses: actions/cache@v1
+        with:
+          path: ~/.npm  # npm caches files in ~/.npm
+          key: node-${{ hashFiles('**/package-lock.json') }}
+
+      - name: Install dependencies
+        run: npm install --no-audit
```
Comment (dhruvkb) [inline .github/workflows/cd.yml]: De-indent by 2 spaces.
```
key:
- val
- val
```

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      run: npm install --prefer-offline --no-audit
+
+    - name: Run build
+      run: npm run build
+
+    - name: Publish package
+      run: npm publish --access public
+      env:
+        NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+
```
Comment (dhruvkb) [inline .github/workflows/cd.yml]: End files with a single newline.

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      uses: actions/cache@v1
+      with:
+        path: ~/.npm  # npm caches files in ~/.npm
+        key: node-${{ hashFiles('**/package-lock.json') }}
+
+    - name: Install dependencies
+      run: npm install --prefer-offline --no-audit
+
+    - name: Run build
+      run: npm run build
```
Comment (dhruvkb) [inline .github/workflows/cd.yml]: Sorry for not noticing this before but this is also not required because building is already a step in deployment.

Comment (dhruvkb) [reply .github/workflows/cd.yml]: And if there is a single task, setup also becomes pointless.

**Review by dhruvkb — CHANGES_REQUESTED** [review]

File: .github/workflows/cd.yml (hunk 1 above)
Comment (SaurabhAgarwala) [reply .github/workflows/cd.yml]: I just checked, actually, both are valid ways for specifying the types of event that will trigger the workflow. And in the [official documentation](https://help.github.com/en/actions/reference/events-that-trigger-workflows#check-run-event-check_run) they have followed this convention. So shall I still replace it?

File: .github/workflows/cd.yml (hunk 3 above)
Comment (SaurabhAgarwala) [reply .github/workflows/cd.yml:86]: I think just the End of File symbol is getting rendered separately. The file ends on line 86 only. I have checked it in my local editor as well.

File: .github/workflows/cd.yml (hunk 4 above)
Comment (SaurabhAgarwala) [reply .github/workflows/cd.yml]: Yes, I agree now. I was also thinking about it while coding this, but then I saw the [ci.yml](https://github.com/creativecommons/vocabulary/blob/master/.github/workflows/ci.yml) file and saw it repeated. But I just realized that there it makes sense as the `setup` job will help in caching and also if the `setup` job itself fails, then other jobs won't run only.

Also, I am removing the caching part as I just [found](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache) that cache can only be accessed in workflows triggered by `push` and `pull_request` events and not in `release` event.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (dhruvkb) [reply .github/workflows/cd.yml]: Just for consistency with the other lists we have used, please do change it.

File: .github/workflows/cd.yml (hunk 4 above)
Comment (dhruvkb) [reply .github/workflows/cd.yml]: Setup was meant to
- act as a fail-fast incase there was a problem with npm dependencies
- cache dependencies for the other three tasks
If there as just one task, these use-cases become irrelevant.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (SaurabhAgarwala) [reply .github/workflows/cd.yml]: Sure, I will change it now

File: .github/workflows/cd.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    - name: Install dependencies
+      run: npm install --no-audit
+
+    - name: Run build
+      run: npm run build
+
+    - name: Publish package
+      run: npm publish --access public
+      env:
+        NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (dhruvkb) [inline .github/workflows/cd.yml]: Publishing to `npm` requires a couple more steps. See [`deploy.sh`](https://github.com/creativecommons/vocabulary/blob/master/deploy.sh). You should run the script itself in this step.

**Review by dhruvkb — CHANGES_REQUESTED** [review]

File: .github/workflows/cd.yml (hunk 5 above)
Comment (SaurabhAgarwala) [reply .github/workflows/cd.yml]: Ok. Actually I was following [this](https://help.github.com/en/actions/language-and-framework-guides/publishing-nodejs-packages#publishing-packages-to-the-npm-registry) and these steps were not there. But, I will update the PR now. Thanks for pointing out.

Also, once this PR is approved, I will update the other one with the required changes.

**Review by dhruvkb — APPROVED** [review]
From the looks of it, it appears that it would work as expected. Just waiting for the `NPM_TOKEN` to be added and then we can test it.

**Comment by kgodey** [comment]
@dhruvkb I added the token.

**Pull request merged**